src/backend/cdb/dispatcher/cdbgang.c

```diff
@@ -714,6 +714,34 @@ getCdbProcessesForQD(int isPrimary)
 	return list;
 }
 
+/*
+ * This function is used for case when we should just free memory (without any
+ * additional allocations).
+ */
+void
+DestroyAllGangs(bool resetSession)
+{
+	if (Gp_role == GP_ROLE_UTILITY)
+		return;
+
+	ELOG_DISPATCHER_DEBUG("DestroyAllGangs");
+
+	/* Destroy CurrentGangCreating before GangContext is reset */
+	if (CurrentGangCreating != NULL)
+	{
+		RecycleGang(CurrentGangCreating, true);
+		CurrentGangCreating = NULL;
+	}
+
+	/* destroy cdb_component_dbs, disconnect all connections with QEs */
+	cdbcomponent_destroyCdbComponents();
+
+	if (resetSession)
+		resetSessionForPrimaryGangLoss();
+
+	ELOG_DISPATCHER_DEBUG("DestroyAllGangs done");
+}
+
```

Comment:
Is it possible to extract the common parts of similar functions `DestroyAllGangs` and `DisconnectAndDestroyAllGangs` into one?